src/awkward/_nplikes/array_module.py

```diff
@@ -146,6 +150,29 @@ def searchsorted(
 
     ############################ manipulation
 
+    def apply_ufunc(
+        self,
+        ufunc: UfuncLike,
+        method: str,
+        args: list[Any],
+        kwargs: dict[str, Any] | None = None,
+    ) -> ArrayLike | tuple[ArrayLike]:
+        # Determine input argument dtypes
+        input_arg_dtypes = [getattr(obj, "dtype", type(obj)) for obj in args]
+        # Resolve these for the given ufunc
+        arg_dtypes = tuple(input_arg_dtypes + [None] * ufunc.nout)
+        resolved_dtypes = ufunc.resolve_dtypes(arg_dtypes)
```

Comment:
`np.broadcast_arrays` will take care of `np.int64(x) → np.array([x, x, ...], np.int64)` conversion that avoids value-based promotion. However, this line (`ufunc.resolve_dtypes`) applies the NEP-50 weak promotion rules for Python scalars, which we also want.